src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs

```diff
@@ -66,32 +68,84 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
             var interfacesToGenerate = interfacesWithDiagnostics.Where(static data => data.Diagnostic is null);
             var invalidTypeDiagnostics = interfacesWithDiagnostics.Where(static data => data.Diagnostic is not null);
 
-            var interfaceContexts = interfacesToGenerate.Select((data, ct) =>
+            var interfaceBaseInfo = interfacesToGenerate.Collect().SelectMany((data, ct) =>
             {
-                // Start at offset 3 as 0-2 are IUnknown.
-                // TODO: Calculate starting offset based on base types.
+                ImmutableArray<(int StartingOffset, ManagedTypeInfo? BaseInterface, bool IsMarkerInterface)>.Builder baseInterfaceInfo = ImmutableArray.CreateBuilder<(int, ManagedTypeInfo?, bool)>(data.Length);
+                // Track the calculated last offsets of the interfaces.
+                // If a type has invalid methods, we'll count them and issue an error when generating code for that
+                // interface.
+                Dictionary<INamedTypeSymbol, int> derivedNextOffset = new(SymbolEqualityComparer.Default);
+                foreach (var iface in data)
+                {
+                    var (starting, baseType, derivedStarting) = CalculateOffsetsForInterface(iface.Symbol);
+                    baseInterfaceInfo.Add((starting, baseType is not null ? ManagedTypeInfo.CreateTypeInfoForTypeSymbol(baseType) : null, starting == derivedStarting));
+                }
+                return baseInterfaceInfo.MoveToImmutable();
+
+                (int Starting, INamedTypeSymbol? BaseType, int DerivedStarting) CalculateOffsetsForInterface(INamedTypeSymbol iface)
```

Comment:
Should we make this static with an extra parameter for the captured derivedNextOffset?